torch/csrc/autograd/engine.cpp

```diff
@@ -72,23 +73,42 @@ struct GraphTask {
   // Notified when a task finishes executing.  Check outstanding_tasks to see
   // if all tasks are done.
   std::condition_variable not_done;
-  const Engine::pre_callback_map& pre_callbacks;
-  const Engine::post_callback_map& post_callbacks;
   std::unordered_map<Function*, InputBuffer> not_ready;
   std::unordered_map<Function*, int> dependencies;
 
+  struct ExecInfo {
+    struct Capture {
+      Capture(int input_idx, int output_idx) : input_idx(input_idx), output_idx(output_idx) {}
+      int input_idx; // within Function inputs
+      int output_idx; // within the output vector of a GraphTask
+    };
+
+    bool should_execute() const {
+      return needed || captures;
+    }
+
+    bool needed = false;
+    std::unique_ptr<std::vector<Capture>> captures;
+  };
+  // Exec info has a bit complicated semantics. If it's empty, it means the task is
+  // run in a "default" mode, which means that all next_functions we encounter should
+  // get executed. If it's not empty, only functions that have an entry and this entry
+  // has needed == True should be executed.
+  std::unordered_map<Function*, ExecInfo> exec_info;
+  std::vector<Variable> captured_vars;
```

Comment:
Is this "captured variables" in the sense of, "free variables in closures which outlive their parent scope"?

Reply:
I don't think so. It's more of "the derivatives that someone requested". We have to capture their values from different places in the graph